### PR TITLE
Feat/wildcard namespaces

### PR DESCRIPTION
**Pull request by Joeavaikath**
Thank you for contributing to Velero!

# Please add a summary of your change

# Does your change fix a particular issue?

Adding design document for issue #1874 

# Please indicate you've done the following:

- [ ] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [ ] [Created a changelog file (`make new-changelog`)](https://velero.io/docs/main/code-standards/#adding-a-changelog) or comment `/kind changelog-not-required` on this PR.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

**Review by weshayutin — APPROVED** [review]

**Comment by kaovilai** [comment]
Thanks for tackling this long-standing issue! The lazy evaluation approach with request-scoped caching is elegant. I have some considerations that might strengthen the design:

## Areas for Consideration

### 1. Namespace Discovery Timing
The design snapshots namespaces at backup start time. It would be helpful to document the expected behavior when namespaces are created/deleted during backup execution. For example:
- If namespace `prod-new` is created after backup starts, will it be included if matching `prod-*`?
- Should this behavior be explicitly documented for user expectations?

### 2. Pattern Complexity
The design mentions only basic glob-style wildcards (`*`). Consider clarifying:
- Will `?` for single character matching be supported (e.g., `prod-?-app`)?
- How will patterns like `*-test-*` behave? 
- Are there any pattern limitations users should be aware of?
- Will invalid patterns be validated at creation time?

### 3. Error Handling
It would be valuable to specify:
- What happens if K8s API calls fail during namespace enumeration? Does the backup fail or proceed with cached/partial data?
- How will users be notified when wildcard patterns match zero namespaces? Will this be a warning, error, or silent success?
- Should there be a `--dry-run` option to preview which namespaces would be included?

### 4. Testing Strategy
The design should outline testing approaches:
- Unit tests covering various wildcard patterns (`*-suffix`, `prefix-*`, `*-middle-*`, edge cases)
- Integration tests with actual K8s clusters
- Performance benchmarks for clusters with varying namespace counts (10, 100, 1000+ namespaces)
- Tests for the thread-safety of concurrent access

### 5. Restore Operations
How will wildcard patterns behave during restore? Will they:
- Use the namespaces captured at backup time?
- Re-evaluate against the current cluster state?
- This is particularly important for disaster recovery scenarios

### 6. Scheduled Backups
For scheduled backups, how do wildcards behave when namespace state changes between runs? This affects:
- Backup consistency over time
- Storage usage if namespace sets vary
- User expectations for scheduled vs manual backups

### 7. Example Usage
It would be helpful to include a concrete example of a backup specification YAML showing how wildcards would be used:

```yaml
apiVersion: velero.io/v1
kind: Backup
metadata:
  name: production-backup
spec:
  includedNamespaces:
  - "prod-*"
  - "production-*"
  excludedNamespaces:
  - "*-test"
  # ... other backup specs
```

These additions would help implementers and users better understand the feature's behavior and limitations. Looking forward to seeing this implemented!

**Comment by Joeavaikath** [comment]
> Thanks for tackling this long-standing issue! The lazy evaluation approach with request-scoped caching is elegant. I have some considerations that might strengthen the design:
> 
> ## Areas for Consideration
> ### 1. Namespace Discovery Timing
> The design snapshots namespaces at backup start time. It would be helpful to document the expected behavior when namespaces are created/deleted during backup execution. For example:
> 
> * If namespace `prod-new` is created after backup starts, will it be included if matching `prod-*`?
> * Should this behavior be explicitly documented for user expectations?
> 
> ### 2. Pattern Complexity
> The design mentions only basic glob-style wildcards (`*`). Consider clarifying:
> 
> * Will `?` for single character matching be supported (e.g., `prod-?-app`)?
> * How will patterns like `*-test-*` behave?
> * Are there any pattern limitations users should be aware of?
> * Will invalid patterns be validated at creation time?
> 
> ### 3. Error Handling
> It would be valuable to specify:
> 
> * What happens if K8s API calls fail during namespace enumeration? Does the backup fail or proceed with cached/partial data?
> * How will users be notified when wildcard patterns match zero namespaces? Will this be a warning, error, or silent success?
> * Should there be a `--dry-run` option to preview which namespaces would be included?
> 
> ### 4. Testing Strategy
> The design should outline testing approaches:
> 
> * Unit tests covering various wildcard patterns (`*-suffix`, `prefix-*`, `*-middle-*`, edge cases)
> * Integration tests with actual K8s clusters
> * Performance benchmarks for clusters with varying namespace counts (10, 100, 1000+ namespaces)
> * Tests for the thread-safety of concurrent access
> 
> ### 5. Restore Operations
> How will wildcard patterns behave during restore? Will they:
> 
> * Use the namespaces captured at backup time?
> * Re-evaluate against the current cluster state?
> * This is particularly important for disaster recovery scenarios
> 
> ### 6. Scheduled Backups
> For scheduled backups, how do wildcards behave when namespace state changes between runs? This affects:
> 
> * Backup consistency over time
> * Storage usage if namespace sets vary
> * User expectations for scheduled vs manual backups
> 
> ### 7. Example Usage
> It would be helpful to include a concrete example of a backup specification YAML showing how wildcards would be used:
> 
> ```yaml
> apiVersion: velero.io/v1
> kind: Backup
> metadata:
>   name: production-backup
> spec:
>   includedNamespaces:
>   - "prod-*"
>   - "production-*"
>   excludedNamespaces:
>   - "*-test"
>   # ... other backup specs
> ```
> 
> These additions would help implementers and users better understand the feature's behavior and limitations. Looking forward to seeing this implemented!

I have added some lines around these issues

File: design/wildcard-namespace-support-design.md [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+    // ExcludeWildcardMatches records the expanded exclude wildcard namespaces
+    // +optional
+    // +nullable
+    ExcludeWildcardMatches []string `json:"excludeWildcardMatches,omitempty"`
+
+    // WildcardResult records the final namespaces after applying wildcard include/exclude logic
+	// +optional
+	// +nullable
+	WildcardResult []string `json:"wildcardResult,omitempty"`
```
Comment (shubham-pampattiwar) [inline design/wildcard-namespace-support-design.md]: NIT: indentation fix

File: design/wildcard-namespace-support-design.md [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+```go
+// BackupStatus captures the current status of a Velero backup.
+type BackupStatus struct {
+    // ... existing fields ...
+    
+    // IncludeWildcardMatches records the expanded include wildcard namespaces
+    // +optional
+    // +nullable
+    IncludeWildcardMatches []string `json:"includeWildcardMatches,omitempty"`
```
Comment (shubham-pampattiwar) [inline design/wildcard-namespace-support-design.md]: Should we encapsulate all these new fields under one single struct ?

File: design/wildcard-namespace-support-design.md [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+The implementation maintains full backward compatibility with existing behavior:
+- The standalone "*" character continues to work as before (include all namespaces, ignore excludes)
+- Existing backup configurations remain unaffected
+
+### Known Limitations
+
+1. **Mixed "*" and wildcard usage**: When using the standalone "*" in includes, the legacy implementation takes precedence and wildcard expansion is skipped. This means you cannot combine "*" (all namespaces) in includes with wildcard patterns in excludes.
+
+2. **Selective exclusion limitation**: The current design does not support selective pattern-based exclusion when including all namespaces via "*".
```
Comment (kaovilai) [inline design/wildcard-namespace-support-design.md]: Is 1. and 2. saying the same thing?

Do you think this can be enhanced in the future?

Comment (Joeavaikath) [reply design/wildcard-namespace-support-design.md]: It's more like 2 different side effects but yes, almost the same

Comment (Joeavaikath) [reply design/wildcard-namespace-support-design.md]: If we try to change how * is treated, then it might have a knock on effect on other tests and other points of code that use that logic
Definitely worth doing since we would want to be able to say back everything up except stuff that matches a certain pattern

**Comment by kaovilai** [comment]
Please fix [dco](https://github.com/vmware-tanzu/velero/pull/9057/checks?check_run_id=47439550628).

> To add your Signed-off-by line to every commit in this branch:
> 
> Ensure you have a local copy of your branch by [checking out the pull request locally via command line](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally).
> In your local branch, run: git rebase HEAD~9 --signoff
> Force push your changes to overwrite the branch: git push --force-with-lease origin feat/wildcard-namespaces

File: design/wildcard-namespace-support-design.md [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+## Backup
+
+The wildcard expansion implementation focuses on two key functions in `pkg/backup/item_collector.go`:
+
+- [`collectNamespaces`](https://github.com/vmware-tanzu/velero/blob/1535afb45e33a3d3820088e4189800a21ba55293/pkg/backup/item_collector.go#L742) - Retrieves all active namespaces and processes include/exclude filters
+- [`getNamespacesToList`](https://github.com/vmware-tanzu/velero/blob/1535afb45e33a3d3820088e4189800a21ba55293/pkg/backup/item_collector.go#L638) - Resolves namespace includes/excludes to final list
+
+The `collectNamespaces` function is the ideal integration point because it:
+- Already retrieves all active namespaces from the cluster
```
Comment (sseago) [inline design/wildcard-namespace-support-design.md]: I was just looking at the collectNamespaces code, since it was referenced in this design. It's actually looking like it's not going to be so helpful for you here, since this function is only used for collecting the namespaces *as kubernetes cluster-scoped resources to back up*. It's not used for determining whether namespace-scoped resources *within this namespace* should be backed up. Basically you will want to modify `collections.IncludesExcludes` instead, since that is used by `collectNamespaces` as well as in the item backupper.

Comment (Joeavaikath) [reply design/wildcard-namespace-support-design.md]: Expanding it earlier means we have to make another call to get all the namespaces, and then check if they're active.

Since collectNamespaces already has the list of active namespaces, we could just expand it there before we pass it forward?
This then updates the NamespaceIncludesExcludes object that's used down the line, so namespace-scoped resources would be backup up, since it now has the expanded namespace list.

I have an implementation that works with this, but if this is not the right place for it can move it earlier up and make a call to get all active namespaces

> namespaces as kubernetes cluster-scoped resources to back up 
> namespace-scoped resources within this namespace should be backed up

As long as we pass the right NamespaceIncludesExcludes object, the namespace-scoped resources should be fine?

**Comment by Joeavaikath** [comment]
Parallel backups need expanded namespaces to begin with, so the current approach will not work since the request would still have wildcard patterns.
Will update the design when this dependency is resolved

edit: new design takes parallel backups into account, so expansion occurs early enough

File: design/wildcard-namespace-support-design.md (hunk 3 above)
Comment (Joeavaikath) [reply design/wildcard-namespace-support-design.md]: This point was a misunderstanding on my part: it does ignore the excludes initially, but just when its grabbing the list via API call.
Eventually ShouldInclude takes the Excludes into account, so it does work normally.

File: design/wildcard-namespace-support-design.md (hunk 4 above)
Comment (Joeavaikath) [reply design/wildcard-namespace-support-design.md]: With this updated design, expansion runs before collectNamespaces and getNamespacesToList
Uses the new struct NamespaceIncludesExcludes's ResolveNamespaceList()

File: design/wildcard-namespace-support-design.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+### Wildcard Expansion Package
+
+New `pkg/util/wildcard/expand.go` package provides:
+
+- `ShouldExpandWildcards()` - Skip expansion for simple "*" case
+- `ExpandWildcards()` - Main expansion function using `github.com/gobwas/glob`
+- Pattern validation rejecting unsupported regex symbols
+
+**Supported patterns**: `*`, `?`, `[abc]`, `{a,b,c}`
```
Comment (kaovilai) [inline design/wildcard-namespace-support-design.md:80]: Is this manually defined? or some kind of standard like https://datatracker.ietf.org/doc/rfc9485/

Comment (Joeavaikath) [reply design/wildcard-namespace-support-design.md:80]: Does not support all regex, so no it's not a standard
Just the few there

**Comment by shubham-pampattiwar** [comment]
@Lyndon-Li @reasonerjt please review :)

**Comment by kaovilai** [comment]
implementation avail at https://github.com/vmware-tanzu/velero/pull/9255

**Review by blackpiglet — APPROVED** [review]
